README.md
[How to Run](#how-to-run-netbots) | [Write a Robot](#how-to-write-a-robot) | [Mechanics](#game-mechanics) | [Configuration](#server-configuration) | [Modules](#module-reference) | [Messages](#messages)

# NetBots

NetBots is a python programming game. The game consists of a number of robots, 4 by default, that battle in an arena. To play, a python program must be written. The program can control the robot's speed and direction, scan for enemy robots and fire exploding shells from its canon. Robots suffer damage if they are in collisions or hit by exploding shells. The game ends when only one robot remains or the time limit is reached. Normally, many games are played in a tournament to determine which robot is the overall winner.

NetBots is inspired by [RobotWar](https://en.wikipedia.org/wiki/RobotWar) from the 1970s. RobotWar has been cloned many times, one popular example is [Crobots](https://en.wikipedia.org/wiki/Crobots). 

The image below is the NetBots viewer. The filled circles are robots and the unfilled circle is an explosion.

**Viewer**

<img src="images/basicgame.png" width="80%">

**Scoreboard**

Below is a sample scoreboard, which is output after each game by the NetBots server.
```

                  ------ Scoreboard ------
               Tournament Time: 1140.929 secs.
                         Games: 1000
             Average Game Time: 1.141 secs.
                         Steps: 569173
          Average Steps / Game: 569.173
                      Run Time: 1140.967 secs.
      Time Processing Messages: 211.796 secs.
                   Messages In: 1953793
                  Messages Out: 1628161
              Messages Dropped: 325632
             Messages / Second: 3424.803
         Time Processing Steps: 15.490 secs.
                Steps / Second: 498.868
                 Time Sleeping: 905.588 secs.
            Average Sleep Time: 0.001589 secs.
     Steps Slower Than stepSec: 184 (0.03%)

                    ---- Score -----  ------ Wins -------  --------- CanonFired ----------                       
              Name      Points     %    Count   AvgHealth    Count   AvgDamage   TotDamage  IP:Port              
 ------------------------------------------------------------------------------------------------------------------
      Super Bot v3        4727  27.9      310       40.59    29729        3.17    94094.14  127.0.0.1:20171      
            Rascal        4411  26.1      296       42.79    24663        3.97    97828.85  127.0.0.1:20182      
           Level 5        4371  25.8      246       36.20    28715        3.50   100594.83  127.0.0.1:20129      
             Oscar        3418  20.2      141       35.94    21832        3.33    72758.46  127.0.0.1:20189      
 ------------------------------------------------------------------------------------------------------------------
```

**Demo Video**

[![NetBots Demo Video](https://img.youtube.com/vi/y2_1gFQjgCo/0.jpg)](https://www.youtube.com/watch?v=y2_1gFQjgCo)

### How is NetBots different?

NetBots differs from RobotWar, and its clones by being real-time and network centric. The server and robots each run in separate processes and can run on the same or separate computers. The server runs at a specific rate (steps/second). Additionally, the server emulates an unreliable network where message (packet) loss is common. Writing programs to deal with the real-time nature and network unreliability provides additional programming challenges.

Finally, NetBots offers other optional challenges for robot logic: [Robot Classes](docs/robot_classes.md), [Obstacles, and Jam Zones](docs/obstacles_and_jam_zones.md).

### NetBots as a Learning Tool

NetBots can be used in a learning environment. Students can be challenged in two ways:

1. Learn to write programs that must interact with a constantly changing real-time environment with limited information and limited control.
2. Learn about networking, the impact of unreliable networks, and synchronous vs asynchronous programming.

See [Proposed Learning Goals](docs/proposed_learning_goals.md) for further ideas.


---


# How To Run NetBots


## Prerequisites


### Python 3.6 or higher

NetBots uses Python 3.6 or higher (tested on python 3.7.3) which can be installed from [https://www.python.org/downloads/](https://www.python.org/downloads/). Only the standard python 3 libraries are required. 

> If multiple versions of python are installed, ensure you are running python 3.6+, not python 3.5 or python 2. The examples in this README use the "python" command assuming python 3.6+ is the default. The command "python3" (Linux) or "py -3" (Windows) may be required to force the correct version.


### NetBots Code

The NetBots code can be cloned with git from: [https://github.com/dbakewel/netbots.git](https://github.com/dbakewel/netbots.git) or downloaded in zip form from: [https://github.com/dbakewel/netbots/archive/master.zip](https://github.com/dbakewel/netbots/archive/master.zip)


## Running the Demo

On windows, **double click "rundemo.bat"** in the root of the NetBots directory.

> If this does not work, open a command window (cmd), cd into the directory containing rundemo.bat and type "rundemo.bat".

The rundemo script will start 6 processes on the local computer: 1 server, 1 viewer, and 4 robots. A default tournament (10 games) will run and then the server will quit. Each process will send its output to its own cmd window. The title of the window indicates what is running it in. Each process can be quit by clicking in the window and pressing "Ctrl-C" (cmd window stays open) or clicking the close box (cmd window closes). Use "Close all windows" in the task bar to quickly quit all processes.

*The demo robots are fairly boring. Robots you write can be much faster and more exciting!*


## Running a Tournament

There are three options available on the NetBots server that are useful for tournaments. The first changes the number of games (**-games**) in the tournament. If robots have similar skills then playing more games will flesh out which robot is best. 

The second option (**-stepsec**) allows you to speed up the NetBots server. Most modern computers can run NetBots 5 times faster (or more) than the default (0.05 sec/step or 20 steps/sec). The server scoreboard will display "Steps Slower Than stepSec", which indicates when it can't keep up with the requested speed. If only a few of steps are slow (<1%) then it will not affect the game. If many steps are slow (>1%) you should stop the server and reduce its target speed.

The final option (**-stepmax**) changes the maximum steps in a game. If most games are ending because the maximum steps is reached than increasing this will give robots more times to demonstrate their skills.

For example, to run a 1000 game tournament at 5 times faster (0.01 sec/step or 100 steps/sec) with a maximum of 2000 steps per game use:

```
python netbots_server -games 1000 -stepsec 0.01 -stepmax 2000
```

## Increasing Netbots Performance

Running tournaments can take a long time. One way to speed up tournaments is to increase the steps per second however this may be limited by hardware. Significant performance increases can be found by installing the binary msgpack package for python. Msgpack is used to encode and decode messages between the robots and server. Running at 2000 steps/second (-stepsec 0.0005) should be possible if: modern hardware is used, binary msgpack is installed, the server and all bots are running on localhost (127.0.0.1), and no viewer is connected. 

To install binary msgpack on Windows use:
```
py -3 -m pip install msgpack-python
```

To install binary msgpack on Linux use:
```
pip3 install msgpack-python
```

To ensure the binary msgpack is being used, look for a log when the server and bots start up:
```
INFO 2020-05-28 23:11:13.115 netbots_ipc.<module>: Using binary python msgpack.
```

## Running Larger Tournaments on Linux

The NetBots server is limited in that it runs a tournament with the same robots in every game. One solution to having more than 4 robots is to increase the number of robots (-bots server) and make the arena larger (-arenasize). While this works it also changes the game dynamics. 

To run a tournament with more than 4 robots but with default settings (4 robots per game and 1000x1000 arena) the divisions_tournament.py script can be used (Linux only). It can run a tournament with a multiple of 4 robots (4, 8, 16, ...) up to 64 total. Robots are put into divisions (4 robots in each). Over consecutive rounds, better robots will move to lower numbered divisions (division 0 being the best). See the rundivisions.sh script for an example of how to run and then customize to meet your needs.


## Running on Separate Computers

By default NetBots only listens on localhost 127.0.0.1 which does not allow messages to be sent or received between computers. To listen on all network interfaces, and allow messages from other computers, use ```-ip 0.0.0.0```. 

> Note, if you want to run NetBots across a network then the ports you choose must be open in the OS and network firewalls for two way UDP traffic. By default, these ports are in the range 20000 - 20020 range but any available UDP ports can be used.

For example:

Assuming:
*   computer 1 has IP address of 192.168.1.10
*   computer 2 has IP address of 192.168.1.30

The server can be run on computer 1 with: 

```
python netbots_server.py -ip 0.0.0.0 -p 20000
```

The server will listen on 127.0.0.1 and 192.168.1.10

A robot can be run on Computer 2 with: 

```
python robot.py -ip 0.0.0.0 -p 20010 -sip 192.168.1.10 -sp 20000
```

A robot can also be run on Computer 1 will the default IP of 127.0.0.01 with: 

```
python robot.py -p 20010 -sp 20000
```

Even though the robots on computer 1 and computer 2 use the same port (20010) they are on separate computers so it works. If you try running two robots on the same port on the same computer you will get an error.

## Command Line Help

The server, viewer, and demo robots all allow some customization with command line switches. Run each with the **-h** switch to display help. For example:

```
D:\netbots\src>python src\netbots_server.py -h
usage: netbots_server.py [-h] [-ip Server_IP] [-p Server_Port]

CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [2.2.0] - 2020-06-16
### Changed
- Significant update to divisional tournaments: 1) Can run more than one division (netbots server) at a time, 2) handles crashed robots, 3) better logging, 4) tunable vars moved to top of script.
- Change server so it runs with whatever bots have joined by the time maxSecsToJoin is reached, as long as 2 robots or more have joined. Server still only allows max botsInGame to join and no more robots can join after game has started.
- The server no longer counts missed steps while waiting for bots to join game.

## [2.1.3] - 2020-05-30
### Changed
- Change scoreboard Missteps from an absolute value to a %.
- Change divisions_tournament.py to perform cross divisions at the start of each round, after the first.
- Fixed bug in divisions_tournament.py so that stderr is directed to stdout correctly.

## [2.1.2] - 2020-05-29
### Changed
- Fixed bug in stop conditions of divisions_tournament.py.

## [2.1.1] - 2020-05-28
### Changed
- Made binary msgpack the default if installed. Updated docs on how to increase NetBots performance with binary msgpack.
- Updated README docs for divisions_tournament.py use on Linux.

## [2.1.0] - 2020-05-27
### Added
- Added Division Tournament runner which can run tournaments with a multiple of 4 robots, up to 64. (4, 8, ..., 64), only on Linux. No documentation has been created yet. For now see code or ask a question on GitHub for help.
- Added server options for: json output (-jsonsb); suppress the output of all but the last scoreboard (-onlylastsb); maximum time server will wait for robots to join game (-maxsecstojoin). These were needed to support divisional tournaments.
- Added "Missteps" to scoreboard. Missteps is the number of steps during which no messages were received from the robot.
- Added file output support to netbots_log via the setLogFile method.
- Added support for binary msgpack module. However this does not work with the viewer and has been left commented out. If not using the viewer then a significant performance increase can be seen with the binary msgpack. See import section of netbots_ipc module for code to uncomment.

### Changed
- Changed all instances of "score board" to "scoreboard".
- Fixed python warning for "is not" being used with a literal. Changed to "!=".

## [2.0.0] - 2019-06-09
### Added
- Robot classes have been significantly improved. See [Robot Classes](docs/robot_classes.md) for details. Robot classes more values, such as explRadius.
- Instant replay added to viewer. Pressing the space bar activates a 7 second long instant replay at the default step speed. Live action will resume after the replay completes.
- Robots in viewer have been updated to indicate most parameters: direction, requested direction, speed, requested speed, last scan, and last fire direction.
- When it quits, NetBots server outputs netbot_ipc stats separated by ip:port.

### Changed
- Advanced collisions is now the default and default hitDamage has been increased to 10. **This means robots will take much more damage from collisions.** Restore old behavior with ```-simplecollisions -hitdamage 1```.
- Demo robot train.py replaced with scaredycat.py. Scaredy Cat demonstrates avoidance of other robots and walls by using scan and location data.
- Other updates to README. Some information from README has been moved under docs/.

## [1.4.0] - 2019-06-02
### Added
- The server now supports advanced collisions (enable with -advancedcollisions option). Hit damage is assigned based on speed and angle of collision.
- Viewer -randcolor option added that uses random color for robots.
- A new robot class system has been implemented. Enable with -allowclasses option on server. There is still more work before classes have correct game mechanics.
- The server now has -scanmaxdistance option which defines the max distance an enemy robot can be detected with scanRequest.
- Team demo robot added. This is an example of using python threading to control two robot from one python script. The robots can share information and work together.
- Data from each robots last scanRequest and fireCanonRequest message are now stored and sent to the viewer. This can be used by later viewer updates.

### Changed
- Demo robots wait longer for server to to reply to joinRequest.
- Robots can no longer send messages with unknown data fields and optional fields are now supported and validated.
- Shells now appear as small arrows in viewer rather than circles.
- Robot names are now limited to max length of 16 characters.
- Fixed bug that did not allow scanReply to return robot further than 1415 distance away.
- Minor updates to README.
- Server no longer mistakenly includes result field in fireCanonReply message.

## [1.3.0] - 2019-05-15
### Added
- Server option -noviewers which disables viewers.
- Server prints command line options on startup.


src/netbots_ipc.py
import socket
import random
import time
import re
import math
import argparse

from netbots_log import log

try:
    import msgpack as umsgpack
    log("Using binary python msgpack.")
except:
    import umsgpack
    log("Using pure python msgpack. Install binary msgpack for better performance.", "WARNING")

"""
**About Messages**

Every message is a python dict type with at least a type attribute (i.e. field) and a 
string value.

For example: { 'type': 'getInfoRequest'}

MsgDef below defines valid types and additional fields that must be included and 
can optionally be included based on type.

Optional fields end in '_o' which marks the fields as optional. The '_o' should not 
appear in the actual message, it is just a marker that the field is optional.
For example, a joinRequest message with optional class field would be:

{'type': 'joinRequest', 'name': 'SuperRobot', 'class': 'heavy'}

All fields have a defined type in the form of <type> or [<type>,min,max].
<type> can be expressed as multiple acceptable types as (<type>,<type>,...)

For fields types of 'str' min and max are the min length and max length of the string.

All Request messages have a corresponding Reply message. The Request is sent to the
server and the server returns the reply message or an Error message.
"""
MsgDef = {
    # msg type              other required msg fields
    'joinRequest': {'name': ['str', 1, 16], 'class_o': ['str', 1, 16]},
    'joinReply': {'conf': 'dict'},

    'getInfoRequest': {},
    'getInfoReply': {'gameNumber': 'int', 'gameStep': 'int', 'health': ['(int,float)', 0, 100], 'points': 'int'},

    'getLocationRequest': {},
    'getLocationReply': {'x': ['(int,float)', 0, 32767], 'y': ['(int,float)', 0, 32767]},

    'getSpeedRequest': {},
    'getSpeedReply': {'requestedSpeed': ['(int,float)', 0, 100], 'currentSpeed': ['(int,float)', 0, 100]},

    'setSpeedRequest': {'requestedSpeed': ['(int,float)', 0, 100]},
    'setSpeedReply': {},

    'getDirectionRequest': {},
    'getDirectionReply': {'requestedDirection': ['(int,float)', 0, math.pi * 2], 'currentDirection': ['(int,float)', 0, math.pi * 2]},

    'setDirectionRequest': {'requestedDirection': ['(int,float)', 0, math.pi * 2]},
    'setDirectionReply': {},

    'getCanonRequest': {},
    'getCanonReply': {'shellInProgress': 'bool'},

    'fireCanonRequest': {'direction': ['(int,float)', 0, math.pi * 2], 'distance': ['(int,float)', 10, 32767]},
    'fireCanonReply': {},

    'scanRequest': {'startRadians': ['(int,float)', 0, math.pi * 2], 'endRadians': ['(int,float)', 0, math.pi * 2]},
    'scanReply': {'distance': ['(int,float)', 0, 32767]},

    'addViewerRequest': {},
    'addViewerReply': {'conf': 'dict'},

    # The msg types below do not have, nor expect, a matching reply
    'viewData': {'state': 'dict', 'bots': 'dict', 'shells': 'dict', 'explosions': 'dict'},
    'viewKeepAlive': {},
    'Error': {'result': 'str'}
}


def isValidMsg(msg):
    """ Returns True if msg is a valid message, otherwise returns false. """

    global MsgDef

    if not isinstance(msg, dict):
        log("Msg is type " + str(type(msg)) + " but must be dict type: " + str(msg), "ERROR")
        return False
    if not 'type' in msg:
        log("Msg does not contain 'type' key: " + str(msg), "ERROR")
        return False

    unvalidedFields = list(msg.keys())
    # type is validated below as part of loop so does not need specific validation.
    unvalidedFields.remove('type')
    # msgId and replyData are always optional and have no specific format. So they are always valid if present.
    if 'msgID' in unvalidedFields:
        unvalidedFields.remove('msgID')
    if 'replyData' in unvalidedFields:
        unvalidedFields.remove('replyData')

    for msgtype, msgspec in MsgDef.items():
        if msgtype == msg['type']:
            for fld, fldspec in msgspec.items():
                if fld.endswith('_o'):
                    # remove magic suffix marking field as optional
                    fld = fld.rstrip('_o')
                    if fld not in msg:
                        # optional field is not present, which is valid.
                        continue
                elif fld not in msg:
                    log("Msg does not contain required '" + fld + "' key: " + str(msg), "ERROR")
                    return False
                if isinstance(fldspec, list):
                    if not isinstance(msg[fld], eval(fldspec[0])):
                        log("Msg '" + fld + "' key has value of type " + str(type(msg[fld])) +
                            " but expected " + fldspec[0] + ": " + str(msg), "ERROR")
                        return False
                    if fldspec[0] is 'str':
                        if len(msg[fld]) < fldspec[1] or len(msg[fld]) > fldspec[2]:
                            log("Msg '" + fld + "' key has a string value " + str(msg[fld]) +
                                " with length out of range [" + str(fldspec[1]) + "," +
                                str(fldspec[2]) + "] : " + str(msg), "ERROR")
                            return False
                    elif msg[fld] < fldspec[1] or msg[fld] > fldspec[2]:
                        log("Msg '" + fld + "' key has a value " + str(msg[fld]) +
                            " which is out of range [" + str(fldspec[1]) + "," +
                            str(fldspec[2]) + "] : " + str(msg), "ERROR")
                        return False
                else:
                    if not isinstance(msg[fld], eval(fldspec)):
                        log("Msg '" + fld + "' key has value of type " + str(type(msg[fld])) +
                            " but expected " + fldspec + ": " + str(msg), "ERROR")
                        return False
                unvalidedFields.remove(fld)
            # All fields defined for message type have now been examined and are valid
            if len(unvalidedFields):
                # message has fields it should not have.
                log("Msg contains field(s) " + str(unvalidedFields) + " which is not defined for message type " + msg['type'] + ": " + str(msg), "ERROR")
                for fld in unvalidedFields:
                    if fld.endswith('_o'):
                        log("Optional message fields should not include '_o' suffix in field name.", "WARNING")
                        break
                return False
            else:
                # message is valid and has no extra fields.
                return True
    log("Msg 'type' key has value '" + str(msg['type']) + "' which is not known: " + str(msg), "ERROR")
    return False


def isValidIP(ip):
    """ Returns True if ip is valid IP address, otherwise returns false. """
    if not isinstance(ip, str):
        log("IP is type " + str(type(ip)) + " but must be type str.", "ERROR")
        return False
    if not re.match(r'^[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}$', ip):
        log("IP address has bad format, expected something like 'int.int.int.int' but got " + ip, "ERROR")
        return False
    return True

# Check IP address format as needed by argparse module.


def argParseCheckIPFormat(ip):
    """ Returns ip if ip is a valid IP address, otherwise raises argparse.ArgumentTypeError exception. """

    if not isValidIP(ip):
        raise argparse.ArgumentTypeError(ip)
    return ip


def isValidPort(p):
    """ Returns True if p is valid port number, otherwise returns false. """

    if not isinstance(p, int):
        log("Port is type " + str(type(p)) + " but must be type int.", "ERROR")
        return False
    if p < 1 or p > 65000:
        log("Port is out of valid range 0-65000: " + str(p), "ERROR")
        return False
    return True


def formatIpPort(ip, port):
    """ Formats ip and port into a single string. eg. 127.168.32.11:20012 """
    return str(ip) + ":" + str(port)


class NetBotSocketException(Exception):
    """Raised by the NetBotSocket class."""
    pass


class NetBotSocket:
    """NetBot Msg filtering and basic reliable send/recv for UDP soket. """

    def __init__(self, sourceIP, sourcePort, destinationIP='127.0.0.1', destinationPort=20000):
        """
        Create and bind UDP socket and bind it to listen on sourceIP and sourcePort.

        sourceIP: IP the socket will listen on. This must be 127.0.0.1 (locahost), 0.0.0.0 (all interfaces), or a valid IP address on the computer.
        sourcePort: port to listen on. This is an integer number.
        destinationIP and destinationPort are stored with setDestinationAddress()


        Returns NetBotSocket object.

        Raises socket related exceptions.
        """

        self.sent = {}  # Number of messages sent to OS socket
        self.recv = {}  # Number of messages recv from OS socket
        self.sendRecvMessageCalls = 0  # Number of calls to sendRecvMessage
        self.sendRecvMessageResends = 0  # Number of resends made by sendRecvMessage
        self.sendRecvMessageTime = 0  # Total time in sendRecvMessage
        self.sendTypes = {}
        self.recvTypes = {}

        self.sendrecvDelay = 0.1

        self.sourceIP = sourceIP
        self.sourcePort = sourcePort
        log("Creating socket with sourceIP=" + sourceIP + ", sourcePort=" + str(sourcePort), "VERBOSE")
        self.s = socket.socket(family=socket.AF_INET, type=socket.SOCK_DGRAM)
        try:
            self.s.bind((sourceIP, sourcePort))
            log("Source Socket Binding Successful. Listening on " + formatIpPort(sourceIP, sourcePort))
        except Exception as e:
            self.s.close()
            self.s = None
            log("Source Socket Binding Failed. The source port may already be in use. Try another port.", "FAILURE")
            raise
        self.s.settimeout(0)
        self.destinationIP = destinationIP
        self.destinationPort = destinationPort
        self.bufferSize = 4096
        random.seed()
        self.msgID = random.randrange(0, 65000, 1)

    def settimeout(self, t):
        self.s.settimeout(t)

    def setDelay(self, delay):
        self.sendrecvDelay = delay

    def getStats(self):
        """ Return str of NetBotSocket stats. """
        output = "\n\n                 ====== Stats ======"

        if self.sendRecvMessageCalls:
            output += \
                "\n     sendRecvMessage Calls: " + str(self.sendRecvMessageCalls) + \
                "\n   sendRecvMessage Resends: " + str(self.sendRecvMessageResends) + \
                "\n  Avg sendRecvMessage Time: " + \
                '%.6f' % (self.sendRecvMessageTime / self.sendRecvMessageCalls) + " secs."

        for src in self.sent.keys():
            output += "\n\n               === To/From: " + src + " ==="\
                "\n             Messages Sent: " + str(self.sent[src]) +\
                "\n             Messages Recv: " + str(self.recv[src])

            if src in self.sendTypes:
                output += "\n\n                Messages Sent by Type"
                for t, c in sorted(self.sendTypes[src].items(), key=lambda x: x[0]):
                    output += "\n" + '%26s' % (t) + ": " + str(c)

            if src in self.recvTypes:
                output += "\n\n                Messages Recv by Type"
                for t, c in sorted(self.recvTypes[src].items(), key=lambda x: x[0]):
                    output += "\n" + '%26s' % (t) + ": " + str(c)

            output += "\n"

        return output

    def setDestinationAddress(self, destinationIP, destinationPort):
        """
        Set default destination used by NetBotSocket send and recv functions when
        destination is not provided.

        Returns no value

        Raises NetBotSocketException exception.
        """

        if not isValidIP(destinationIP) or not isValidPort(destinationPort):
            raise NetBotSocketException("Bad IP or Port Provided.")

        self.destinationIP = destinationIP
        self.destinationPort = destinationPort

    def serialize(self, msg):
        return umsgpack.packb(msg, use_bin_type=True)

    def deserialize(self, b):
        return umsgpack.unpackb(b, raw=False)

    def sendMessage(self, msg, destinationIP=None, destinationPort=None, packedAndChecked=False):
        """
        Sends msg to destinationIP:destinationPort and then returns immediately.
        sendMessage is considered asynchronous because it does not wait for a
        reply message and returns no value. Therefore there is no indication if
        msg will be received by the destination.

        msg must be a valid message (see Messages below). Raises
        NetBotSocketException exception if the msg does not have a valid format.

        If destinationIP or destinationPort is not provided then the default will
        be used (see setDestinationAddress()).

        If packedAndChecked is True then msg is assumed to already be serialized
        and no other checks will be done.

        """

        if destinationIP is None:
            destinationIP = self.destinationIP

        if destinationPort is None:
            destinationPort = self.destinationPort

        if not packedAndChecked:
            if not isValidMsg(msg):
                raise NetBotSocketException("Could not send because msg is not valid format.")
            if not isValidIP(destinationIP):
                raise NetBotSocketException("Could not send because destinationIP is not valid format.")
            if not isValidPort(destinationPort):
                raise NetBotSocketException("Could not send because destinationPort is not valid format.")

            # Convert data from python objects to network binary format
            networkbytes = self.serialize(msg)
        else:
            networkbytes = msg

        log("Sending msg to " + destinationIP + ":" + str(destinationPort) +
            " len=" + str(len(networkbytes)) + " bytes " + str(msg), "DEBUG")
        self.s.sendto(networkbytes, (destinationIP, destinationPort))

        dest = formatIpPort(destinationIP, destinationPort)
        if dest in self.sent:
            self.sent[dest] += 1
        else:
            self.sent[dest] = 1

        if not packedAndChecked:
            msgtype = msg['type']
        else:
            msgtype = "Serialized"

        if dest not in self.sendTypes:
            self.sendTypes[dest] = {}
        if msgtype in self.sendTypes[dest]:
            self.sendTypes[dest][msgtype] += 1
        else:
            self.sendTypes[dest][msgtype] = 1

    def recvMessage(self):
        """
        Check the socket receive buffer and returns message, ip, and port only
        if a valid message is immediately ready to receive. recvMessage is
        considered asynchronous because it will not wait for a message to arrive
        before raising an exception.

        Returns msg, ip, port.
              msg: valid message (see Messages below)
              ip: IP address of the sender
              port: port of the sender

        If the reply is an “Error” message then it will be returned just like
        any other message. No exception will be raised.

        If msg is not a valid message (see Messages below) then raises
        NetBotSocketException.

        Immediately raises NetBotSocketException if the receive buffer is empty.

        Note, the text above assumes the socket timeout is set to 0
        (non-blocking), which is the default in NetBotSocket.

        """
        try:
            bytesAddressPair = self.s.recvfrom(self.bufferSize)
            # Convert data from network binary format to python objects
            msg = self.deserialize(bytesAddressPair[0])
            ip = bytesAddressPair[1][0]
            port = bytesAddressPair[1][1]
            log("Received msg from " + ip + ":" + str(port) + " len=" +
                str(len(bytesAddressPair[0])) + " bytes " + str(msg), "DEBUG")

            src = formatIpPort(ip, port)
            if src in self.recv:
                self.recv[src] += 1
            else:
                self.recv[src] = 1

            if src not in self.recvTypes:
                self.recvTypes[src] = {}
            if msg['type'] in self.recvTypes[src]:
                self.recvTypes[src][msg['type']] += 1
            else:
                self.recvTypes[src][msg['type']] = 1

        except (BlockingIOError, socket.timeout):
            # There was no data in the receive buffer.
            raise NetBotSocketException("Receive buffer empty.")
        except (ConnectionResetError):
            # Windows raises this when it gets back an ICMP destination unreachable packet
            log("The destination ip:port returned ICMP destination unreachable. Is the destination running?", "WARNING")
            raise NetBotSocketException(
                "The destination ip:port returned ICMP destination unreachable. Is the destination running?")

        if not isValidMsg(msg):
            raise NetBotSocketException("Received message invalid format.")

        # If we get a joinReply then use the server conf to tune our send delay in sendRecvMessage()
        if msg['type'] == 'joinReply':
            self.setDelay(msg['conf']['stepSec'] * 2)

        return msg, ip, port

    def sendRecvMessage(self, msg, destinationIP=None, destinationPort=None,
                        retries=10, delay=None, delayMultiplier=1.2):
        """
        Sends msg to destinationIP:destinationPort and then returns the reply.
        sendRecvMessage is considered synchronous because it will not return
        until and unless a reply is received. Programmers can this of this much
        like a normal function call.

        msg must be a valid message (see Messages below)

        If destinationIP or destinationPort is not provided then the default will
        be used (see setDestinationAddress()).

        If the reply is an “Error” message then a NetBotSocketException exception
        will be raised.

        If no reply is received then the message will be sent again (retried) in
        case it was dropped by the network. If the maximum number of retries is
        reached then a NetBotSocketException exception will be raised.

        Raises NetBotSocketException exception if the msg does not hae a valid format.
        """

        startTime = time.perf_counter()
        self.sendRecvMessageCalls += 1

        if destinationIP is None:
            destinationIP = self.destinationIP

        if destinationPort is None:
            destinationPort = self.destinationPort

        if delay:
            nextDelay = delay
        else:
            nextDelay = self.sendrecvDelay

        remaining = retries

        self.msgID = self.msgID + 1
        if self.msgID > 65000:
            self.msgID = 0

        msg['msgID'] = self.msgID

        gotReply = False
        sendMessage = 0
        while remaining != 0 and gotReply == False:
            if sendMessage <= time.perf_counter():
                self.sendMessage(msg, destinationIP, destinationPort)
                if sendMessage != 0:
                    self.sendRecvMessageResends += 1
                remaining = remaining - 1
                sendMessage = time.perf_counter() + nextDelay
                self.s.settimeout(nextDelay)
                nextDelay = nextDelay * delayMultiplier

            try:
                replyMsg, ip, port = self.recvMessage()
            except NetBotSocketException as e:
                # We didn't get anything from the buffer or it was an invalid message.
                ip = None

            if ip is not None:
                # if the message is the one we are looking for.
                if ip == destinationIP and port == destinationPort and \
                        isinstance(replyMsg, dict) and \
                        'msgID' in replyMsg and replyMsg['msgID'] == msg['msgID']:
                    gotReply = True

        self.s.settimeout(0)

        if not gotReply:
            log("Raising Exception NetBotSocketException because failed to get valid respose after " + str(retries) +
                " retries with delay = " + str(delay) + " and delayMultiplier = " + str(delayMultiplier), "VERBOSE")
            raise NetBotSocketException("Failed to get valid respose.")

        if replyMsg['type'] == "Error":
            log("Raising Exception NetBotSocketException because reply message, with correct msgID was of type Error.",
                "VERBOSE")
            raise NetBotSocketException("Received Error Message: " + replyMsg['result'])

        del replyMsg['msgID']

        self.sendRecvMessageTime += time.perf_counter() - startTime
        return replyMsg
